### PR TITLE
Bugfix gradients

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
The `more_strategies` PR introduced several custom gradients, but we never noticed they were broken under some conditions.

**Description of the Change:**
- Gradients are now cast to the backend array type
- This PR also adds the numpy option to the bargmann methods

**Benefits:**
Gradients work

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: mrmustard/lab/abstract/state.py [hunk 1]
```diff
@@ -526,15 +526,19 @@ def __getitem__(self, item) -> State:
         self._modes = item
         return self
 
-    def bargmann(self) -> Optional[tuple[ComplexMatrix, ComplexVector, complex]]:
-        r"""Returns the Bargmann representation of the state."""
+    def bargmann(self, numpy=False) -> Optional[tuple[ComplexMatrix, ComplexVector, complex]]:
```
Comment (elib20) [inline mrmustard/lab/abstract/state.py:529]: Is there any function that needs numpy=True? I don't see one in this PR. We have several mrmustard functions that return tensorflow objects, then usually it's on the user to convert it to numpy if they need for their purposes.

File: tests/test_training/test_opt.py [hunk 2]
```diff
@@ -457,3 +457,14 @@ def cost_fn():
 
     assert np.allclose(bsgate.theta.value, 0.1, atol=0.01)
     assert np.allclose(bsgate.phi.value, 0.2, atol=0.01)
+
+
+def test_squeezing_grad_from_fock():
```
Comment (elib20) [inline tests/test_training/test_opt.py:462]: Should we test the other gates that have custom gradients too?

Comment (elib20) [reply tests/test_training/test_opt.py:462]: This worked.

`from mrmustard.training import Optimizer, Parametrized
from mrmustard.lab import *

sq = SqueezedVacuum(r=1.0,r_trainable=True)

def cost_fn():
    return -((sq &Fock(1)) >> BSgate(0.5) << (Vacuum(1)&Fock(1)))

opt = Optimizer(euclidean_lr=0.05)
opt.minimize(cost_fn, by_optimizing=[sq], max_steps=100)`

Comment (elib20) [reply tests/test_training/test_opt.py:462]: This did not.

`from mrmustard.training import Optimizer, Parametrized
from mrmustard.lab import *

disp = Dgate(x=1.0,y=1.0, x_trainable=True,y_trainable=True)

def cost_fn():
    return -(Fock(2)>> disp << Vacuum(1))

opt = Optimizer(euclidean_lr=0.05)
opt.minimize(cost_fn, by_optimizing=[disp], max_steps=100)`

Comment (elib20) [reply tests/test_training/test_opt.py:462]: This worked.

`from mrmustard.training import Optimizer, Parametrized
from mrmustard.lab import *

bs = BSgate(theta=0.5,theta_trainable=True)

def cost_fn():
    return -((Fock(2,cutoffs=[3]) &Fock(1,cutoffs=[3])) >> bs << (Fock(2,cutoffs=[3]) &Fock(1,cutoffs=[3])))

opt = Optimizer(euclidean_lr=0.05)
opt.minimize(cost_fn, by_optimizing=[bs], max_steps=100)`

**Review by elib20 — COMMENTED** [review]
Thanks for patching that so quickly!

I am still having issues with the Dgate.

**Comment by sduquemesa** [comment]
Why no changelog for this change?

File: mrmustard/lab/abstract/state.py (hunk 1 above)
Comment (ziofil) [reply mrmustard/lab/abstract/state.py:529]: Not in this PR no, but I keep needing to transform the triple into numpy, so I just added the option here.

**Comment by ziofil** [comment]
> Why no changelog for this change?

Because it's not fixing a bug in a previous version, so it wouldn't make sense to claim a bug was fixed in the upcoming release

File: tests/test_training/test_opt.py (hunk 2 above)
Comment (ziofil) [reply tests/test_training/test_opt.py:462]: caught it!

**Review by elib20 — APPROVED** [review]
Nice!
